Adding GetHandleInformation and SetHandleInformation

exvito opened the pull request:
Following #65 here is my first take at adding these APIs.

Preliminary work:
* I did read the README.rst file and setup my environment as documented.
* In master, commit c29bf1b5dc8966d1b0c84533ea6e78e44accbff6, some tests are failing:  `FAILED (SKIP=8, errors=4)` - looks related to the fact that I don't have Visual Studio, they all `raise DistutilsPlatformError("Unable to find vcvarsall.bat")`
  * tests.test_core.test_dist.TestLoad
  * tests.test_core.test_dist.TestImportPath
  * tests.test_core.test_dist.TestCompile
  * tests.test_core.test_dist.TestCompile
* Other than that, the 9 tests in tests/test_kernel32/test_handle.py are passing (even though one of those is outputting `close failed in file object destructor: IOError: [Errno 9] Bad file descriptor`.

Regarding my code:
* I tried to follow existing code layout and conventions.
* Handle test count now 16 and passing: I'd appreciate your review and improvement recommendations.
* I'm not sure how to add / use / make the associated constants available: I see several constants declared in pywincffi/core/cdefs/headers/constants.h, but don't understand:
   * Why their respective values are missing and `...` is used instead.
   * How to use them within pywincffi code (not needed currently).
   * How to expose them to pywincffi users.

Additionally, I created a one-off test program which makes effective use of the APIs. It binds a socket, forks a child, closes the socket, rebinds the socket to the same address and stops the child. Without the SetHandleInformation call to clear the socket handle's information the rebinding should fail on Python < 3.4 (which is my case).

Here it is, for reference:
```python
# -------------------------------------------------------------------

from __future__ import print_function

import sys
import subprocess
import socket

from pywincffi import kernel32


# -------------------------------------------------------------------

class Handle(object):

    import cffi as _cffi
    _ffi = _cffi.FFI()

    def __init__(self, integer):
        self._handle = Handle._ffi.cast('HANDLE', integer)
        self._addr = Handle._ffi.cast('unsigned long', self._handle)

    def __repr__(self):
        return '<Handle 0x%x>' % self._addr

    @property
    def value(self):
        return self._handle


# -------------------------------------------------------------------

if __name__ == '__main__':

    bind_addr = ('127.0.0.1', 10001)

    s = socket.socket()
    s_handle = Handle(s.fileno()) # fileno() is socket handle on Win
    s.bind(bind_addr)
    s_addr = s.getsockname()
    print('bound socket %s to %r' % (s_handle, s_addr))

    inherit = kernel32.GetHandleInformation(s_handle.value)
    print('socket handle inheritance flag: %s' % inherit)

    kernel32.SetHandleInformation(s_handle.value, 1, 0)
    print('cleared socket handle inheritance flag')

    inherit = kernel32.GetHandleInformation(s_handle.value)
    print('socket handle inheritance flag: %s' % inherit)

    p = subprocess.Popen(
        args=[sys.executable],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE
    )
    print('child PID:', p.pid)

    s.close()
    print('closed socket')

    s = socket.socket()
    s_handle = Handle(s.fileno())
    try:
        s.bind(bind_addr)
        s_addr = s.getsockname()
        print('re-bound socket %s to %r' % (s_handle, s_addr))
    except Exception as e:
        print('re-binding socket failed:', e)

    # stop the child
    p.stdin.close()
    print('child exit code:', p.wait())


# -------------------------------------------------------------------
```

Thanks for your review and feedback.

opalmer commented:
I'll make most of my comments in-line with the code, but here's some replies to some of your specific questions:

> I'm not sure how to add / use / make the associated constants available: I see several constants declared in pywincffi/core/cdefs/headers/constants.h, but don't understand:
> Why their respective values are missing and ... is used instead.

I'm using `...` so the compiler can [fill in the gaps for us](http://cffi.readthedocs.org/en/latest/cdef.html#letting-the-c-compiler-fill-the-gaps).  

When it comes to Windows constants code in Python I see a lot of this kind of stuff:

``` python
FILE_ATTRIBUTE_ENCRYPTED = 16384 
# or 
FILE_ATTRIBUTE_ENCRYPTED = 0x4000
```

In pywincffi I've instead done the below in `pywincffi/core/cdefs/headers/constants.h`:

``` C
#define FILE_ATTRIBUTE_ENCRYPTED ...
```

There's some distinct advantages to this:
- You can't create a constant which does not exist which eliminates typos.  If you do have a typo in the name it results in an error at compile time which is more effective at spotting issues than code review would be.
- You can't create a properly named constant with the wrong value.  The compiler adds the value from what it knows is the correct value.  Just like the name of the constant it would be fairly easy to miss the in a review.
- Once compiled the constant cannot be dynamically modified, deleted, etc.  This prevents someone from accidentally messing with a constant which could be an issue for either their own code or internally within pywincffi.

> How to use them within pywincffi code (not needed currently).

There's a few places in the code base which are using the constants already but basically it works like this:

``` python
from pywincffi.core import dist
ffi, library = dist.load()
print(library.PIPE_WAIT)
```

Looking at the code you have you probably should start using constants by name now.  There's a couple of reasons for this:
- By using the constants internally, even in tests, you're making it harder to make simple mistakes like passing in the wrong int.
- By using the constants internally you're forced to create them in `pywincffi/core/cdefs/headers/constants.h`.  That way when someone needs them they'll be available before they're requested.  This kind of follows the 'eating your own dog food' philosophy if you're familiar with the term.

I basically make it a habit to always include any constants referenced inside of the msdn documentation when I'm implementing a function.  Side note, I'm working on #67 to document some of these practices for the future.

> How to expose them to pywincffi users.

They're already exposed.  An end user will just need to load the library first using `dist.load()`.

> Additionally, I created a one-off test program which makes effective use of the APIs. It binds a socket, forks a child, closes the socket, rebinds the socket to the same address and stops the child. Without the SetHandleInformation call to clear the socket handle's information the rebinding should fail on Python < 3.4 (which is my case).

Could you turn this into a test? The code you have is basically what I'd normally expect to see in a test which asserts that the behavior itself works as expected.  A good example of this is the `TestLockFileEx` test case which uses `LockFileEx` to lock a file then try and access the same file in another process.  While it's more complex than just making sure the function runs it's closer to a real world use case which also means the tests can be used as examples to consumers of pywincffi.

**Post Review Edit**

Finished the review and overall it looks great, thanks!

Most of the comments are around style issues, consistency, etc but I mostly found it to be pretty consistent with the rest of the project. I still think you should take the code you have above and add it in the form of a test though. 

So to summarize, here's the TODOs:
- [x] Add the code you have above as a test.
- [x] Address the comments from the review or reply if you have questions.
- [x] Get the PR checks passing (see 'Details' next to each line item). (#68 will fix one of the problems with the build)
- [ ] Add an entry to `docs/changelog.rst` under the 'latest' section once #68 merges

opalmer commented on pywincffi/kernel32/handle.py:
```diff
@@ -120,3 +120,50 @@ def WaitForSingleObject(hHandle, dwMilliseconds):
+    :param handle hObject:
+        A handle to an object whose information is to be set.
+
+    :param DWORD dwMask:
```
The type should be the expected type for Python.  So rather than `DWORD` this should be `int`.

opalmer commented on pywincffi/kernel32/handle.py:
```diff
@@ -120,3 +120,50 @@ def WaitForSingleObject(hHandle, dwMilliseconds):
+    :param DWORD dwMask:
+        A mask that specifies the bit flags to be changed.
+
+    :param DWORD dwFlags:
```
The type should be the expected type for Python.  So rather than `DWORD` this should be `int`.

opalmer commented on pywincffi/kernel32/handle.py:
```diff
@@ -120,3 +120,50 @@ def WaitForSingleObject(hHandle, dwMilliseconds):
+    input_check("dwFlags", dwFlags, integer_types)
+    _, library = dist.load()
+
+    code = library.SetHandleInformation(hObject, dwMask, dwFlags)
```
Instead of 

``` python
code = library.SetHandleInformation(hObject, dwMask, dwFlags)
```

Could you do:

``` python
code = library.SetHandleInformation(
    hObject, 
    ffi.cast("DWORD", dwMask), 
    ffi.cast("DWORD", dwFlags)
)
```

In #67, I added some reasoning behind this:
https://github.com/opalmer/pywincffi/commit/f6006d2aea78470e4715b7415f86aecdaa6742d1#diff-0f010e46bc45ff3689c2205c91cd62c3R232

opalmer commented on pywincffi/kernel32/handle.py:
```diff
@@ -120,3 +120,50 @@ def WaitForSingleObject(hHandle, dwMilliseconds):
+
+    .. seealso::
+
+        https://msdn.microsoft.com/en-us/library/windows/desktop/ms724935(v=vs.85).aspx
```
For consistency, could you match this url to the others:

```
.. seealso::

   https://msdn.microsoft.com/en-us/library/ms724935

```

The above url is shorter which means it easier to read.  It also allows us to go back and search/replace all of the msdn references in the future since they'll be identical.

opalmer commented on pywincffi/kernel32/handle.py:
```diff
@@ -120,3 +120,50 @@ def WaitForSingleObject(hHandle, dwMilliseconds):
+
+    .. seealso::
+
+        https://msdn.microsoft.com/en-us/library/windows/desktop/ms724329(v=vs.85).aspx
```
For consistency, could you match this url to the others:

```
.. seealso::

   https://msdn.microsoft.com/en-us/library/ms724329

```

The above url is shorter which means it easier to read.  It also allows us to go back and search/replace all of the msdn references in the future since they'll be identical.

exvito replied:
Question: both MSDN and Google searches link to the longer URLs I used. What's the source of these much nicer ones?

opalmer replied:
I've seen it in a couple of training sessions and I played around with the url structure some too.  Microsoft's robots.txt and sitemap tends to reference the direct urls, which is part of the reason Google's bot sees the full url.  I was a bit apprehensive about using the shorter urls but after investigating it some, they're not redirects:

```
> curl -vL https://msdn.microsoft.com/en-us/library/ms724329
...
> GET /en-us/library/ms724329 HTTP/1.1
> Host: msdn.microsoft.com
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 200 OK
...
```

Fun fact, this works too and might be something I update to later (I assume Microsoft's application will select the correct language): https://msdn.microsoft.com/library/ms724329

opalmer commented on pywincffi/kernel32/handle.py:
```diff
@@ -120,3 +120,50 @@ def WaitForSingleObject(hHandle, dwMilliseconds):
+    input_check("hObject", hObject, Enums.HANDLE)
+    ffi, library = dist.load()
+
+    lpdwFlags = ffi.new('LPDWORD')
```
I'd prefer we remain consistent with regards to `'` vs `"` (ie, almost always use `'`).  I've tried to document the reasoning behind this here: https://github.com/opalmer/pywincffi/commit/f9946a7224351fce178bde550f8b23b7ff5be7d9#diff-dc4147bbe08af630a0a380407d3097c2R7

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -1,13 +1,15 @@
+import sys
```
nit: import order should be alphabetical so I'd expect:

``` python
import os
import socket
import sys
import tempfile
```

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+    """
+    Tests for :func:`pywincffi.kernel32.GetHandleInformation`
+    """
+    def _expected_inheritance(self):
```
I'd simplify this function to:

``` python
return sys.version_info[0:2] < (3, 4)
```

exvito replied:
This will return a bool. We need a 0 or 1 integer.
Given that "explicit is better than implicit", maybe `return 1 if sys.version_info[:2] < (3,4) else 0`.
What do you say?

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+
+    def test_get_handle_info_file(self):
+        # can't use mkstemp: not inheritable on Python < 3.4
+        dirname = tempfile.mkdtemp()
```
Two comments here:
1. `dirname` is a function name in the standard library so we should name this something else to eliminate confusion and limit conflicts should we use the function in the future.  I'd call this `tempdir` or something else descriptive. 
2. Don't forget to cleanup after your tests.

``` python
self.addCleanup(shutil.rmtree, tempdir, ignore_errors=True)
```

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+
+    def test_get_handle_info_socket(self):
+        ffi, _ = dist.load()
+        s = socket.socket()
```
Variable names should be descriptive `sock` would be fine here.

Also I suggest you use `try: finally:` to ensure `close()` is called so the cleanup functions shouldn't fail if parts of the test do

``` python
try:
    # everything after the socket.socket() and before close()
finally:
    sock.close()
```

exvito replied:
For consistency, I'll go the addCleanup route, unless you recommend against it for some reason I am unaware of.

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+
+    def test_set_handle_info_socket_inherit(self):
+        ffi, _ = dist.load()
+        s = socket.socket()
```
Variable names should be descriptive `sock` would be fine here.

Also I suggest you use `try: finally:` to ensure `close()` is called so the cleanup functions shouldn't fail if parts of the test do

``` python
try:
    # everything after the socket.socket() and before close()
finally:
    sock.close()
```

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+
+    def test_set_handle_info_socket_noinherit(self):
+        ffi, _ = dist.load()
+        s = socket.socket()
```
Variable names should be descriptive `sock` would be fine here.

Also I suggest you use `try: finally:` to ensure `close()` is called so the cleanup functions shouldn't fail if parts of the test do

``` python
try:
    # everything after the socket.socket() and before close()
finally:
    sock.close()
```

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+        test_file.write("data")
+        file_handle = handle_from_file(test_file)
+
+        SetHandleInformation(file_handle, 1, 0)
```
We should use a constant rather than hard coding integers for constants.  Since the msdn documentation for `SetHandleInformation` references two constants I suggest you add those to constants.h and then use `HANDLE_FLAG_INHERIT` here.

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+        s = socket.socket()
+        socket_handle = ffi.cast('void *', s.fileno())
+
+        SetHandleInformation(socket_handle, 1, 1)
```
We should use a constant rather than hard coding integers for constants.  Since the msdn documentation for `SetHandleInformation` references two constants I suggest you add those to constants.h and then use `HANDLE_FLAG_INHERIT` here.

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+        s = socket.socket()
+        socket_handle = ffi.cast('void *', s.fileno())
+
+        SetHandleInformation(socket_handle, 1, 0)
```
We should use a constant rather than hard coding integers for constants.  Since the msdn documentation for `SetHandleInformation` references two constants I suggest you add those to constants.h and then use `HANDLE_FLAG_INHERIT` here.

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+        test_file.write("data")
+        file_handle = handle_from_file(test_file)
+
+        SetHandleInformation(file_handle, 1, 1)
```
We should use a constant rather than hard coding integers for constants.  Since the msdn documentation for `SetHandleInformation` references two constants I suggest you add those to constants.h and then use `HANDLE_FLAG_INHERIT` here.

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+        SetHandleInformation(file_handle, 1, 1)
+
+        test_file.close()
+        os.unlink(filename)
```
Cleanups should be done using `self.addCleanup(...)` and `addCleanup` should be called right after the resource is created.

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+    Tests for :func:`pywincffi.kernel32.SetHandleInformation`
+    """
+    def test_set_handle_info_file_inherit(self):
+        dirname = tempfile.mkdtemp()
```
Two comments here:
1. `dirname` is a function name in the standard library so we should name this something else to eliminate confusion and limit conflicts should we use the function in the future.  I'd call this `tempdir` or something else descriptive. 
2. Don't forget to cleanup after your tests.

``` python
self.addCleanup(shutil.rmtree, tempdir, ignore_errors=True)
```

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+        os.rmdir(dirname)
+
+    def test_set_handle_info_file_noinherit(self):
+        dirname = tempfile.mkdtemp()
```
Two comments here:
1. `dirname` is a function name in the standard library so we should name this something else to eliminate confusion and limit conflicts should we use the function in the future.  I'd call this `tempdir` or something else descriptive. 
2. Don't forget to cleanup after your tests.

``` python
self.addCleanup(shutil.rmtree, tempdir, ignore_errors=True)
```

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+        else:
+            return 1
+
+    def test_get_handle_info_stdin(self):
```
This test is failing on Python 3.4:

https://ci.appveyor.com/project/opalmer/pywincffi/build/726

exvito replied:
Went re-reading PEP 0446. It seems to omit details regarding STDIN's default inheritance -- acceptable given the fact that the title says "Make newly created ...".

We can either throw this test away, hardcode the expected result to 1 or just validate it is either 1 or 0.

I went with the latter: would you do it differently?

opalmer replied:
Nope, looks good to me thanks.

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+        SetHandleInformation(file_handle, 1, 0)
+
+        test_file.close()
+        os.unlink(filename)
```
Don't call cleanup functions yourself because it can introduce errors if it fails (see notes above around `addCleanup`).

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+        expected = self._expected_inheritance()
+        self.assertEqual(inherit, expected)
+        test_file.close()
+        os.unlink(filename)
```
Don't call cleanup functions yourself because it can introduce errors if it fails (see notes above around `addCleanup`).

opalmer commented on pywincffi/kernel32/handle.py:
```diff
@@ -120,3 +120,50 @@ def WaitForSingleObject(hHandle, dwMilliseconds):
+        https://msdn.microsoft.com/en-us/library/windows/desktop/ms724329(v=vs.85).aspx
+
+    :param handle hObject:
+        A handle to an object whose information is to be retrieved.
```
Could you add some documentation for the return value here?  Something like

``` python
def GetHandleInformation(...):
    """
    ...

    :returns:
        Returns the a bit set containing the handle properties or 0.
    """
```

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+        filename = os.path.join(dirname, "test_file")
+        test_file = open(filename, "w")
+        test_file.write("data")
+        file_handle = handle_from_file(test_file)
```
Since you're not using a `with open(...)` I suggest you use `try: finally:` to ensure `close()` is called so the cleanup functions shouldn't fail if parts of the test do

``` python
try:
    # everything after the open and before close()
finally:
    test_file.close()
```

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+    def test_set_handle_info_file_inherit(self):
+        dirname = tempfile.mkdtemp()
+        filename = os.path.join(dirname, "test_file")
+        test_file = open(filename, "w")
```
Since you're not using a `with open(...)` I suggest you use `try: finally:` to ensure `close()` is called so the cleanup functions shouldn't fail if parts of the test do

``` python
try:
    # everything after the open and before close()
finally:
    test_file.close()
```

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+    def test_set_handle_info_file_noinherit(self):
+        dirname = tempfile.mkdtemp()
+        filename = os.path.join(dirname, "test_file")
+        test_file = open(filename, "w")
```
Since you're not using a `with open(...)` I suggest you use `try: finally:` to ensure `close()` is called so the cleanup functions shouldn't fail if parts of the test do

``` python
try:
    # everything after the open and before close()
finally:
    test_file.close()
```

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+    def test_set_handle_info_socket_inherit(self):
+        ffi, _ = dist.load()
+        s = socket.socket()
+        socket_handle = ffi.cast('void *', s.fileno())
```
nit: Use `"` instead of `'` (nit because these are tests)

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+    def test_set_handle_info_socket_noinherit(self):
+        ffi, _ = dist.load()
+        s = socket.socket()
+        socket_handle = ffi.cast('void *', s.fileno())
```
nit: Use `"` instead of `'` (nit because these are tests)

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,95 @@ def test_process_dies_before_timeout(self):
+    def test_get_handle_info_socket(self):
+        ffi, _ = dist.load()
+        s = socket.socket()
+        socket_handle = ffi.cast('void *', s.fileno())
```
nit: Use `"` instead of `'` (nit because these are tests)

exvito commented:
Update:
- Addressed your review comments.
- Some might need further review, pending on your feedback.

Next:
- I'll integrate my sample code into a new test and a variation of it testing file inheritance.
- Maybe I'll add a few more Get/Set simpler tests.
- Update `docs/changelog.rst` once #68 is merged.

Side-question:
Thanks for the constant related feedback. I browsed the CFFI docs and I'm half puzzled half amazed.
Given that the constants are C preprocessor defines which end up as unnamed constants in the final binaries and the fact that my system does not have any actual headers from where that information can be extracted, how in the world does CFFI manage to get the associated values? (and it does, indeed!)

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,86 @@ def test_process_dies_before_timeout(self):
+        stdin_handle = GetStdHandle(library.STD_INPUT_HANDLE)
+        handle_flags = GetHandleInformation(stdin_handle)
+        inherit = handle_flags & library.HANDLE_FLAG_INHERIT
+        expected = (0, 1)
```
I'd modify this a bit so we're testing specific values depending on the version of Python.  That will make it clear what the expectation is per Python version.

exvito replied:
I investigated this:
- PEP 0446 says nothing about STDIN: it only applies to "newly created" files.
- The first commit of the test did just that: it tested for specific values depending on Python version (non-inheritable from 3.4 onwards, inheritable otherwise). It failed under Python 3.4 precisely because because of that, STDIN came out as inheritable in every Python version on AppVeyor.
- Microsoft docs seem to indicate that STDIN is always inheritable even though I could not find an explicit statement of that fact. The reasoning seems to make sense, though: if a child process is created and its STDIN is not redirected, it will use the parent's STDIN which needs to be inheritable.

Having said this, the best I could come up with was confirming that we were getting a valid value out of GetHandleInformation for STDIN.

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +122,86 @@ def test_process_dies_before_timeout(self):
+        self.assertEqual(inherit, expected)
+
+
+class TestSetHandleInformation(TestCase):
```
Global comment for all tests in this test case.  Would you have objections to calling `GetHandleInformation` after each call to `SetHandleInformation` and testing the output?  

Technically speaking from a Windows perspective so long as the call to `SetHandleInformation` succeeds (since we're checking the error code) it should be fine.  But if it were me, I'd add the extra test just to be sure because it ensures that we can modify and then read data from the same handle with two different API calls.

This kind of goes along with your "Maybe I'll add a few more Get/Set simpler tests." maybe?

exvito replied:
That's what I meant with "Maybe I'll add a few more Get/Set simpler tests". The existing ones are more unit-test like.

The new test case would include more complex integration-like tests:
- Set followed by Get.
- The socket bind, fork, re-bind (converting my manual test code).
- An equivalent one for a file, probably along the lines of: create file, fork, rename file.

opalmer commented on pywincffi/kernel32/handle.py:
```diff
@@ -120,3 +120,58 @@ def WaitForSingleObject(hHandle, dwMilliseconds):
+    code = library.SetHandleInformation(
+        hObject,
+        ffi.cast("DWORD", dwMask),
+        ffi.cast("DWORD", dwFlags),
```
Don't need the `,` at the end.

exvito replied:
This comes as a habit: when declaring lists, dicts, etc. with per-line elements, I always leave the last `,` to simplify adding new elements or moving them around.

I'll take it out if you prefer.

opalmer replied:
Looking at PEP8 again, they don't have a preference there.  For consistency though, let's remove it.  I do get the argument for leaving it to simplify adding new elements but in the case of `SetHandleInformation` there won't be another element to add.

opalmer commented:
> Thanks for the constant related feedback. I browsed the CFFI docs and I'm half puzzled half amazed.
> Given that the constants are C preprocessor defines which end up as unnamed constants in the final binaries and the fact that my system does not have any actual headers from where that information can be extracted, how in the world does CFFI manage to get the associated values? (and it does, indeed!)

It's something cffi does itself I believe rather than the C preprocessor.  I found this out when I tried to do this in constants.h which I know exists on my system as a constant so should work:

``` C
// For the moment, we can't define this here.  When cffi
// parses the header this returns -1 and cffi seems to
// only handle positive integers right now.
//#define INVALID_HANDLE_VALUE ...
```

A lot of the work is done by [cparser.py](https://bitbucket.org/cffi/cffi/src/aca84fd9f4e491b0cdbc0efb536c49d8fbfd8b4a/cffi/cparser.py?at=default&fileviewer=file-view-default) but I'm honestly not exactly sure how it pull the values in since cparser.py is just used to read the constants you want to pull values for.  My guess is cffi either asks another library or knows how to find the value for the constant itself when we call `verify()`.

Now I'm a little curious myself...I might go looking for this (I ask their mailing list)

**Review Comments**

Looking good, nothing major to change in the latest patch set just have the last remaining bullets in your 'Next' list it looks like.

exvito commented:
Updated the tests and included two integration ones:
- Open file, spawn child, rename file in parent.
- Bind socket, spawn child, re-bind socket.

Could not get `test_file_rename_after_fork` to cleanup after itself if the rename fails (commenting out SetHandleInformation, for example). Apparently, shutils.rmtree is not doing is thing for some reason. Can you see why?

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +124,162 @@ def test_process_dies_before_timeout(self):
+            # prevent file_handle inheritance
+            SetHandleInformation(file_handle, library.HANDLE_FLAG_INHERIT, 0)
+            # spawn child while test_file is open
+            p = subprocess.Popen(
```
This should work just as well, though I'l admit I have not tested it against your test so I could be missing something, and should keep the process alive for a short period of time and guarantee it's cleaned up after the test:

``` python
p = subprocess.Popen(
    args=[sys.executable, "-c", "import time; time.sleep(5)"],
    stdin=subprocess.PIPE,
    stdout=subprocess.PIPE,
    stderr=subprocess.PIPE
)
self.addCleanup(p.terminate())
```

exvito replied:
I prefer the approach I took, which is totally synchronised: launching a Python sub-process controlling its IO and, later, closing its STDIN and waiting for its exit.

The `time.sleep` approach becomes asynchronous and nothing in the test enforces the child process to be running while the tested operations are executed.

opalmer commented on tests/test_kernel32/test_handle.py:
```diff
@@ -120,3 +124,162 @@ def test_process_dies_before_timeout(self):
+            sock_handle = ffi.cast("void *", sock.fileno())
+            SetHandleInformation(sock_handle, library.HANDLE_FLAG_INHERIT, 0)
+            # spawn child while sock is bound
+            p = subprocess.Popen(
```
Same comment as above with regards to `subprocess.Popen`

opalmer commented:
> Could not get test_file_rename_after_fork to cleanup after itself if the rename fails (commenting out SetHandleInformation, for example). Apparently, shutils.rmtree is not doing is thing for some reason. Can you see why?

Hmm not sure and I'm not seeing anything in the test output on AppVeyor that points to an error either.  The removal could be failing because a file is still which when combined with `ignore_errors=True` results in the cleanup being skipped essentially.  If we don't ignore the error you should be able to see the cause but that could also cause the test to fail too.

exvito commented:
Your sub-process suggestion with the `self.addCleanup` gave me an idea about why the cleanup I mentioned before was failing. I'll review the test.

exvito commented:
Update:
- Fixed the test that was not cleaning up after itself in the failure case (the child would still be running; now the test ensures the child has terminated before allowing cleanup to run).
- Renamed the tests: "fork" is POSIX terminology, went with the more generic "spawn".

opalmer commented:
Looks good, thanks!  I'll go ahead and merge this tomorrow unless you have more you'd like to do.

exvito commented:
Sounds great but, per your suggestion, `docs/changelog.rst` should be updated in this PR, correct?
Shall I rebase my branch to do that? Will I be able to push it here after rebasing? (talking from inexperience in rebasing within a public context)

opalmer commented:
You should be able push here after a rebase so long as it's the master branch.  That said it would probably be easier to merge this PR and after that I can update #60 with the info before doing a release.

exvito commented:
Sounds good. I'll stand by on this PR's merge and hold on #60.
